tests/std/tests/Dev10_860421_deque_push_back_pop_front/test.cpp

```diff
@@ -47,3 +46,190 @@ void test_391805() {
+
+class NonswappableMovable {
+public:
+    NonswappableMovable() = default;
+    NonswappableMovable(NonswappableMovable&& other) noexcept : payload{exchange(other.payload, -1)} {}
+    NonswappableMovable(const NonswappableMovable&) = default;
+
+    explicit NonswappableMovable(int n) noexcept : payload{n} {}
+    explicit NonswappableMovable(ThrowingConstruction) {
+        throw UniqueError{};
+    }
+
+    NonswappableMovable& operator=(NonswappableMovable&& other) noexcept {
+        payload = exchange(other.payload, -1);
+        return *this;
+    }
+    NonswappableMovable& operator=(const NonswappableMovable&) = default;
+
+#if _HAS_CXX20
+    friend bool operator==(const NonswappableMovable&, const NonswappableMovable&) = default;
+#else // ^^^ C++20 or later / C++17 or earlier
+    friend bool operator==(const NonswappableMovable& lhs, const NonswappableMovable& rhs) noexcept {
+        return lhs.payload == rhs.payload;
+    }
+
+    friend bool operator!=(const NonswappableMovable& lhs, const NonswappableMovable& rhs) noexcept {
+        return lhs.payload != rhs.payload;
+    }
+#endif // ^^^ C++17 or earlier ^^^
+
+    friend void swap(NonswappableMovable, NonswappableMovable) = delete;
+
+private:
+    int payload = -1;
+};
+
+#if _HAS_CXX17
+static_assert(!is_swappable_v<NonswappableMovable>);
+#endif // _HAS_CXX17
+
+void test_exception_safety_for_nonswappable_movable() {
+    using Diff = deque<NonswappableMovable>::difference_type;
+
+    deque<NonswappableMovable> d;
+    for (int i = 0; i < 10; ++i) {
+        d.emplace_back(i);
+    }
+
+    {
+        auto it = d.emplace(d.begin() + Diff{3}, 42);
+        assert(it == d.begin() + Diff{3});
+        assert(d[3] == NonswappableMovable{42});
+    }
+    {
+        auto it = d.emplace(d.end() - Diff{3}, 1729);
+        assert(it == d.end() - Diff{4});
+        assert(d[d.size() - 4] == NonswappableMovable{1729});
+    }
+
+    const auto d_orig = d;
+    try {
+        d.emplace_front(ThrowingConstruction{});
+        assert(false);
+    } catch (const UniqueError&) {
+    }
+    assert(d == d_orig);
+
+    try {
+        d.emplace_back(ThrowingConstruction{});
+        assert(false);
+    } catch (const UniqueError&) {
+    }
+    assert(d == d_orig);
+
+    try {
+        d.emplace(d.begin() + Diff{2}, ThrowingConstruction{});
+        assert(false);
+    } catch (const UniqueError&) {
+    }
+    assert(d == d_orig);
+
+    try {
+        d.emplace(d.end() - Diff{2}, ThrowingConstruction{});
+        assert(false);
+    } catch (const UniqueError&) {
+    }
+    assert(d == d_orig);
+}
+
+class ThrowingMovable {
+public:
+    ThrowingMovable() = default;
+    ThrowingMovable(ThrowingMovable&&) {
+        throw UniqueError{};
+    }
+    ThrowingMovable(const ThrowingMovable&) noexcept = default;
+
+    explicit ThrowingMovable(int n) noexcept : payload{n} {}
+
+    ThrowingMovable& operator=(ThrowingMovable&& other) {
+        throw UniqueError{};
+    }

```

Comment:
This is failing due to an unreferenced parameter warning. As a reminder, please always run any modified test directories.

Reply:
😿I ran the modified test locally in a project but I forgot to compile it with the options for tests. I should do this in a more proper way, sorry.